sqlite_data_norm.c

```diff
@@ -414,6 +414,321 @@ sqlite_fdw_data_norm_float(sqlite3_context* context, int argc, sqlite3_value** a
+	}
+	len = sqlite3_value_int(len_arg);
+	if (vt != SQLITE_INTEGER || (len !=MACADDR_LEN && len !=MACADDR8_LEN))
+	{
+		ereport(ERROR,
+		(errcode(ERRCODE_FDW_INVALID_DATA_TYPE),
+		 errmsg("internal mac deparse error or SQLite input have not 'int' affinity")));
+		return;
+	}
+	if (len == MACADDR_LEN )
+	{
+		char	   *result = (char *) palloc(MACADDR_LEN * 4);
+		sqlite3_int64	value = sqlite3_value_int64(arg);
+		snprintf(result, MACADDR_LEN * 4, "%02x:%02x:%02x:%02x:%02x:%02x",
+  			(unsigned char)((value >> (CHAR_BIT * 5)) & 0xFF),
+  			(unsigned char)((value >> (CHAR_BIT * 4)) & 0xFF),
+  			(unsigned char)((value >> (CHAR_BIT * 3)) & 0xFF),
+  			(unsigned char)((value >> (CHAR_BIT * 2)) & 0xFF),
+  			(unsigned char)((value >> (CHAR_BIT * 1)) & 0xFF),
+  			(unsigned char)((value >> (CHAR_BIT * 0)) & 0xFF)
+  			);
+  		sqlite3_result_text(context, (char*)result, MACADDR_LEN * 4, SQLITE_TRANSIENT);
+	}
+	if (len == MACADDR8_LEN )
+	{
+		char	   *result = (char *) palloc(MACADDR8_LEN * 4);
+		sqlite3_int64	value = sqlite3_value_int64(arg);
+		snprintf(result, MACADDR8_LEN * 4, "%02x:%02x:%02x:%02x:%02x:%02x:%02x:%02x",
+  			(unsigned char)((value >> (CHAR_BIT * 7)) & 0xFF),
+  			(unsigned char)((value >> (CHAR_BIT * 6)) & 0xFF),
+  			(unsigned char)((value >> (CHAR_BIT * 5)) & 0xFF),
+  			(unsigned char)((value >> (CHAR_BIT * 4)) & 0xFF),
+  			(unsigned char)((value >> (CHAR_BIT * 3)) & 0xFF),
+  			(unsigned char)((value >> (CHAR_BIT * 2)) & 0xFF),
+  			(unsigned char)((value >> (CHAR_BIT * 1)) & 0xFF),
+  			(unsigned char)((value >> (CHAR_BIT * 0)) & 0xFF)
+  			);
+  		sqlite3_result_text(context, (char*)result, MACADDR8_LEN * 4, SQLITE_TRANSIENT);
+	}
+}
+
+/*
+ * Attempt to parse a zero-terminated input string zs into a int64
+ * as 6 bytes MAC address. Return 1 if ok, or 0 if the input string is not
+ * parsable.
+ */
+static int
+sqlite_fdw_macaddr6_int (const unsigned char* s, sqlite_uint64*	i)
+{
+	 int		 a,
+				 b,
+				 c,
+				 d,
+				 e,
+				 f;
+	 char		junk[2];
+	 int		 count;
+	 const char* str = (const char*)s;
+
+	 /* %1s matches iff there is trailing non-whitespace garbage */
+
+	 count = sscanf(str, "%x:%x:%x:%x:%x:%x%1s",
+					&a, &b, &c, &d, &e, &f, junk);
+	 if (count != MACADDR_LEN)
+		 count = sscanf(str, "%x-%x-%x-%x-%x-%x%1s",
+						&a, &b, &c, &d, &e, &f, junk);
+	 if (count != MACADDR_LEN)
+		 count = sscanf(str, "%2x%2x%2x:%2x%2x%2x%1s",
+						&a, &b, &c, &d, &e, &f, junk);
+	 if (count != MACADDR_LEN)
+		 count = sscanf(str, "%2x%2x%2x-%2x%2x%2x%1s",
+						&a, &b, &c, &d, &e, &f, junk);
+	 if (count != MACADDR_LEN)
+		 count = sscanf(str, "%2x%2x.%2x%2x.%2x%2x%1s",
+						&a, &b, &c, &d, &e, &f, junk);
+	 if (count != MACADDR_LEN)
+		 count = sscanf(str, "%2x%2x-%2x%2x-%2x%2x%1s",
+						&a, &b, &c, &d, &e, &f, junk);
+	 if (count != MACADDR_LEN)
+		 count = sscanf(str, "%2x%2x%2x%2x%2x%2x%1s",
+						&a, &b, &c, &d, &e, &f, junk);
```

Comment:
> Could you define and use a macro like this?

Very good idea, thanks! Please confirm current implementation after changes, @son-phamngoc.

Reply:
Because we separated into 2 macros `PARSE_MACADDR` and `PARSE_MACADDR8`, I think it is unnecessary to pass `len` into the macro.  
Could you update to be like this?  
```
#define PARSE_MACADDR(len, format) \
if (count != MACADDR_LEN)  \
	count = sscanf(str, format, &a, &b, &c, &d, &e, &f, junk);
```
```
PARSE_MACADDR("%x:%x:%x:%x:%x:%x%1s");
PARSE_MACADDR("%x-%x-%x-%x-%x-%x%1s");
```

Reply:
Changed. Please verify @son-phamngoc !

Reply:
@mkgrgis Do not hard-code `6` and `8` like that. Please use `MACADDR_LEN` and `MACADDR8_LEN`.

Reply:
Fixed in https://github.com/pgspider/sqlite_fdw/pull/101/commits/1ea5b2552c61680fdcb8588669fcb5dc34e27b9d , @son-phamngoc

Reply:
Thanks. Confirmed. Please help me to resolve this comment.